### PR TITLE
[docker] reduce image size

### PR DESCRIPTION
**Pull request by KrisThielemans**
This should reduce the docker image size.
I introduce the `REMOVE_BUILD_FILES` variable (defaulting to 1)

### CODE REVIEW

**Comment by KrisThielemans** [comment]
Failure as we cannot run the ctests anymore, as the `builds` directory was removed. So somehow we have to get rid of that after running the `ctest`. However, as far as I understand, due to the layer system of docker, this isn't so obvious. If we first create the layer with all files present, and then in another delete them, the actual image will not be smaller, see e.g. https://stackoverflow.com/a/47166456/15030207

Note that that post also says that changing perimissions of a file effectively duplicates the file in the image (i.e. it occurs once in each layer), so this seems to say that we should combine the following 3 `RUN` statements into 1
https://github.com/SyneRBI/SIRF-SuperBuild/blob/79cc17a0b11228a3cb40951434e8d9643824735d/docker/Dockerfile#L145-L149

**Comment by KrisThielemans** [comment]
I'll remove the "docker push" stuff again as it's more complicated than I thought (of course!), so will have to leave this for later.

**Comment by KrisThielemans** [comment]
"Fixes" #763 as in this PR, we no longer use `docker run` for the tests. (I haven't checked what happens when running something by hand, but I suggest to close #763 anyway when we merge this PR as it is probably outdated).

**Comment by KrisThielemans** [comment]
Current size of the saved "latest" image is [3371393508](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4045836761/jobs/6957940311#step:6:520). A relatively recent built gave [3562318577](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4013991896/jobs/6894138360#step:9:558). So thi sseems to have saved us only 190MB. Strange.

**Comment by KrisThielemans** [comment]
Merging the `chmod INSTALL` with the build in one `RUN` brought it down to [3154396233](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4050170546/jobs/6967342121#step:6:491), which is another 210MB. Amazingly this is more than the remobal of the `builds`. Hard to understand...

**Comment by KrisThielemans** [comment]
@paskino @lauramurgatroyd this should be ready now. Can you have a look? Only modest decrease sadly (~12% after the `gzip`), but it's something.

I've now added a `docker history` to the GHA to get some more info. Not sure yet if that works, but all the rest is fine from my perspective

**Comment by KrisThielemans** [comment]
Here's what [`docker history` says for layer size on `latest`](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:32)
```
IMAGE          CREATED             CREATED BY                                      SIZE      COMMENT
e51ecc923a69   1 second ago        /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B        
0c6dd2e29d75   9 seconds ago       /bin/sh -c #(nop)  USER root                    0B        
6c49249a084c   18 seconds ago      /bin/sh -c #(nop)  ENV DEBIAN_FRONTEND=         0B        
f15404b2cf1a   27 seconds ago      |14 BUILD_FLAGS= -DCMAKE_BUILD_TYPE=Release …   2.05GB    
04ab6b0c106c   About an hour ago   /bin/sh -c #(nop)  ARG EXTRA_BUILD_FLAGS=-DB…   0B        
17db5a541c82   About an hour ago   /bin/sh -c #(nop)  ARG BUILD_FLAGS= -DCMAKE_…   0B        
ea5068560ef4   About an hour ago   /bin/sh -c #(nop)  ARG NUM_PARALLEL_BUILDS=2    0B        
6ce09cb93ce0   About an hour ago   /bin/sh -c #(nop)  ARG RUN_CTEST=1              0B        
424206115dbd   About an hour ago   /bin/sh -c #(nop)  ARG REMOVE_BUILD_FILES=1     0B        
3c93f065a0ba   About an hour ago   /bin/sh -c #(nop)  ARG SIRF_SB_TAG=master       0B        
be9c65bb6ae9   About an hour ago   /bin/sh -c #(nop)  ARG SIRF_SB_URL=https://g…   0B        
2d78a095490b   About an hour ago   /bin/sh -c #(nop) COPY file:eadc056f025a03d5…   1.69kB    
4f5747daaa99   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   24B       
42eb3c235b18   About an hour ago   /bin/sh -c #(nop)  USER jovyan                  0B        
e3141b69813f   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   0B        
8af4a8161679   About an hour ago   /bin/sh -c #(nop) COPY dir:1d1066138c335283a…   0B        
b762178f909f   About an hour ago   /bin/sh -c #(nop)  USER root                    0B        
3a2122dbed79   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   3.16GB    
ed60875f2d8a   About an hour ago   /bin/sh -c #(nop) COPY file:338e5109c1536908…   2.08kB    
a48552a7003a   About an hour ago   /bin/sh -c #(nop) COPY file:3d17ccae80465e04…   711B      
9e612b517b3c   About an hour ago   /bin/sh -c #(nop) COPY file:c8ccc50d165b79d0…   439B      
048928bd56b6   About an hour ago   /bin/sh -c #(nop)  USER jovyan:users            0B        
102105d17c90   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   1.36kB    
c461ecd25ebe   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   1.55kB    
34a00212f908   About an hour ago   /bin/sh -c #(nop) COPY file:83a4abd22e52e270…   173B      
f75c73b45cf5   About an hour ago   /bin/sh -c #(nop) COPY file:9aca86d46e53dafb…   1.38kB    
c8ed941cf0ac   About an hour ago   /bin/sh -c #(nop) WORKDIR /home/jovyan          0B        
a68551ba0813   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   6.66kB    
d4fd67da4931   About an hour ago   /bin/sh -c #(nop) COPY file:83a4abd22e52e270…   173B      
582ddf2223ee   About an hour ago   /bin/sh -c #(nop) COPY file:9aca86d46e53dafb…   1.38kB    
4538b8df5077   About an hour ago   |7 HOME=/home/jovyan NB_GID=100 NB_GROUP=use…   0B        
0747c315[32](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:33)7e   About an hour ago   /bin/sh -c #(nop)  ENV DEBIAN_FRONTEND=nonin…   0B        
1478550510f0   About an hour ago   /bin/sh -c #(nop)  ARG PYTHON_EXECUTABLE=min…   0B        
177eba3020e2   About an hour ago   /bin/sh -c #(nop)  ARG PYTHON_INSTALL_DIR=/o…   0B        
f634c2d3d42a   About an hour ago   /bin/sh -c #(nop)  ARG HOME=/home/jovyan        0B        
b5b5[33](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:34)631fdd   About an hour ago   /bin/sh -c #(nop)  ARG NB_GID=100               0B        
ba3fba2561ea   About an hour ago   /bin/sh -c #(nop)  ARG NB_UID=22222             0B        
524d5df0b0d8   About an hour ago   /bin/sh -c #(nop)  ARG NB_GROUP=users           0B        
d058d5a3b004   About an hour ago   /bin/sh -c #(nop)  ARG NB_USER=jovyan           0B        
f9145398e886   About an hour ago   /bin/sh -c #(nop)  LABEL description=CCP Syn…   0B        
da889a696b4f   About an hour ago   /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B        
23869e280f67   About an hour ago   /bin/sh -c #(nop)  ENTRYPOINT ["/usr/local/b…   0B        
0102a92b0629   About an hour ago   /bin/sh -c #(nop) COPY file:ae0c6057765c65a2…   1.25kB    
6c110a17712b   About an hour ago   /bin/sh -c #(nop) WORKDIR /home-away            0B        
e309245d2b91   About an hour ago   /bin/sh -c #(nop)  ENV DEBIAN_FRONTEND=         0B        
55c4f350a471   About an hour ago   /bin/sh -c apt-get update -qq && apt-get ins…   4.04MB    
639748dfb7fd   About an hour ago   /bin/sh -c [ -e /usr/bin/X ] || ln -s /usr/b…   0B        
aadbf23cc8ea   About an hour ago   /bin/sh -c mkdir -p /usr/share/X11/xkb          0B        
6ab4cbe54560   About an hour ago   /bin/sh -c apt-get update -qq && apt-get ins…   3.58MB    
ebef1ebe42e4   About an hour ago   /bin/sh -c rm build_system-ubuntu.sh            0B        
fa5960cabbee   About an hour ago   /bin/sh -c bash build_system-ubuntu.sh          520MB     
84335e6b8efe   About an hour ago   /bin/sh -c #(nop) COPY file:5e8ea6627c13a8a7…   2.7kB     
10cb01ee14f0   About an hour ago   /bin/sh -c rm build_gadgetron-ubuntu.sh         0B        
7330fbebfdf6   About an hour ago   /bin/sh -c bash build_gadgetron-ubuntu.sh       696MB     
f463698dfaee   About an hour ago   /bin/sh -c #(nop) COPY file:54730316430f2f85…   923B      
8a4a985f843a   About an hour ago   /bin/sh -c rm build_python-ubuntu.sh            0B        
41d519fb962d   About an hour ago   /bin/sh -c bash build_python-ubuntu.sh          79.9MB    
a4bcc2275397   About an hour ago   /bin/sh -c #(nop) COPY file:e98c1516b079672e…   146B      
dcda5a791303   About an hour ago   /bin/sh -c rm build_essential-ubuntu.sh         0B        
e924a9eb361e   About an hour ago   /bin/sh -c bash build_essential-ubuntu.sh       414MB     
bd0af62e983f   About an hour ago   /bin/sh -c #(nop) COPY file:3d07df126b5e85ce…   811B      
a466e6c1d7df   About an hour ago   /bin/sh -c #(nop)  ENV LANGUAGE=en_GB:en        0B        
697288a10473   About an hour ago   /bin/sh -c #(nop)  ENV LANG=en_GB.UTF-8         0B        
1e5bcf4539cf   About an hour ago   /bin/sh -c rm raw-ubuntu.sh                     0B        
468e714a62e4   About an hour ago   /bin/sh -c bash raw-ubuntu.sh                   20.9MB    
c3c8bbed5f9e   About an hour ago   /bin/sh -c #(nop) COPY file:db2c6a5fef4e30ab…   [34](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:35)2B      
c83e5b14494b   About an hour ago   /bin/sh -c bash update_nvidia_keys.sh && rm …   52.6MB    
325d8015a14b   About an hour ago   /bin/sh -c #(nop) COPY file:3d7c931bcc2d607a…   444B      
8a[37](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:38)2f2[44](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:45)9db   About an hour ago   /bin/sh -c #(nop)  ENV DEBIAN_FRONTEND=nonin…   0B        
8079b[45](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:46)5a817   About an hour ago   /bin/sh -c #(nop)  USER root                    0B        
16793d96b034   About an hour ago   /bin/sh -c #(nop)  LABEL author.name=Casper …   0B        
6b7dfa7e8fdb   7 weeks ago         /bin/sh -c #(nop)  CMD ["bash"]                 0B        
<missing>      7 weeks ago         /bin/sh -c #(nop) ADD file:[48](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:49)1dd2da6de71[52](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:53)52…   [77](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4051663920/jobs/6970679134#step:7:78).8MB
```
Biggest ones (the rest is **much** smaller):
- `user_sirf-ubuntu.sh` 2GB (essentially, build and install the SB). In a previous run I checked the sizes of the build subdirs and 
[found](https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/4050170546/jobs/6967342121#step:4:25869)
    - `INSTALL` 0.5GB
    - `sources` 1.3 GB (checking locally it seems that about 0.5GB of that is git history).
-  [user_python-ubuntu.s](https://github.com/SyneRBI/SIRF-SuperBuild/blob/12ddf53862ce2b599f69c3fbf2670c4e6f11f892/docker/user_python-ubuntu.sh) 3.2GB (essentially install miniconda/mamba and `requirements.yml`)

So, resolving #722  would allow us to remove `sources` as well, which should therefore save us about 1.3GB of around 7GB. (and then compressed). I have no idea what the biggest culprit is for the Python dependencies.

**Comment by KrisThielemans** [comment]
Note that using `gzip -9` for the saved docker images reduced the `latest` artefact from 3154396233 to 3143269798, at the expense of an extra 15 min run-time. So I guess I'll undo that one!

**Comment by KrisThielemans** [comment]
@paskino @lauramurgatroyd this is now finished. Have a look if you can. Merge if you're happy (and the tests worked!)

File: docker/README.md [hunk 1]
```diff
@@ -312,9 +312,10 @@ docker rmi <IMAGEID>
 
 ## Notes
 
-- Tests of SIRF and other installed packages can be run as follows:
+- Since SIRF 3.5, by default the build files are removed on the docker image. This can be changed by [setting an environment variable](./DocForDevelopers.md#Useful-environment-variables).
```
Comment (paskino) [inline docker/README.md:315]: 3.5?

**Review by paskino — APPROVED** [review]

**Comment by paskino** [comment]
I also understood that we should limit the number of layers, i.e. we should put together as many commands as possible as you did here

https://github.com/SyneRBI/SIRF-SuperBuild/blob/7a8901ffe2e5a22142bf379cf9c6b901a234ef61/docker/Dockerfile#L154-L156

This has been pointed out to me by the STFC cloud team too.

**Comment by KrisThielemans** [comment]
> I also understood that we should limit the number of layers, i.e. we should put together as many commands as possible as you did here

yes, but I didn't see any other major opportunities (as far as I recall)

**Comment by KrisThielemans** [comment]
@paskino this is working now.  I suggest to merge (not squash-merge please)

**Pull request merged**